src/main/java/at/hannibal2/skyhanni/features/misc/update/ChangelogViewer.kt

```diff
@@ -0,0 +1,362 @@
+                        )
+                    } else {
+                        ErrorManager.skyHanniError(
+                            "Idk how you ended up here",
+                            "changelog" to changelogList,
+                            "transformed" to this,
+                            "show beta" to shouldShowBeta
+                        )
+                    }
+                )
+            }),
+            height,
+            velocity = 12.0,
+            horizontalAlign = RenderUtils.HorizontalAlignment.CENTER,
+            scrollValue = changelogScroll,
+            button = 0
+        )
+    }
+
+    private fun makeChangeLogToRenderable(
+        it: Map<String, List<String>>,
+        width: Int,
+    ) = it.mapNotNull { (key, value) ->
+        if (!showTechnicalDetails && key == "§l§9Technical Details") {
+            return@mapNotNull null
+        }
+        value.map {
+            Renderable.wrappedString(
+                it, width
+            )
+        }
+    }.flatten()
+
+    private data class VersionTag(
+        val first: Int,
+        val second: Int,
+        val third: Int,
+        val fourth: Int,
+        val isBeta: Boolean,
+    ) : Comparable<VersionTag> {
+
+        constructor(l: List<Int>, beta: Boolean) : this(
+            l.getOrNull(0) ?: -1, l.getOrNull(1) ?: -1, l.getOrNull(2) ?: -1, l.getOrNull(3) ?: -1, beta
+        )
+
+        override operator fun compareTo(other: VersionTag): Int {
+            val first = first.compareTo(other.first)
+            if (first != 0) return first
+            val second = second.compareTo(other.second)
+            if (second != 0) return second
+            val beta = -isBeta.compareTo(other.isBeta)
+            if (beta != 0) return beta
+            val third = third.compareTo(other.third)
+            if (third != 0) return third
+            return fourth.compareTo(other.fourth)
+        }
+
+        override fun toString(): String {
+            return if (isBeta) {
+                "$first" + if (second == -1) " Beta" else ".$second" + if (third == -1) " Beta" else " Beta $third" + if (fourth == -1) "" else ".$fourth"
+            } else {
+                "$first" + if (second == -1) "" else ".$second" + if (third == -1) "" else ".$third" + if (fourth == -1) "" else ".$fourth"
+            }
+        }
+
+        fun isValid() = first != -1
+    }
+
+    /** Inclusive for both borders */
+    private fun VersionTag.isInBetween(current: VersionTag, target: VersionTag): Boolean {
+        if (this > target) return false
+        if (this < current) return false
+        if (this == current) return true
+        return true
+    }
+
+    private fun String.toVersionTag(): VersionTag {
+        val split = this.split('.')
+        val ints = split.filter { it.isInt() }.map { it.toInt() }
+        return VersionTag(ints, split.contains("Beta"))
+    }
+
+    private fun getChangelog(currentVersion: VersionTag, targetVersion: VersionTag) {
```

Comment:
I dont think doing this in the mod is the right choice, I think this should be done in a workflow and similar to the regexes the minecraft formatted changelog should be uploaded there. 

People already have issues with 429 errors on github I dont think calling more github apis is a good idea when it can be automatically uploaded to the repo that has to be downloaded on startup anyway.

Reply:
TLDR: I would want to stay at the approach of using the github api but inform the user about the rate limit.

This is completely optional for player so when the get rate limited the would not care that much. I can add a extra case for the rate limited state, so that we are transparent. Also I don't like to add a changelog to the repo each update since that while increase the repo to stupid amounts (over a long period of time. Which about 1,8KB per day (uncompressed) and that brings the repo since to 0,5MB over a year just from changelogs and the repo sit at the moment at 0,38MB for all things for comperisson). Sure there is no effektiv limit on size on the repo but I don't want to grow the repo infidelity.

Also most of the user will one use 1 token (from 60 per hour). So I don't see it that rate limiting is a issue here, since it isn't automatic like everything else that uses tokens. And the calls to the github api are completely cached (get's cleared after 30min of not using the feature) meaning even if the open it multiple times the only use 1 token. The only case where the use multiple tokens is when it the current version is really old and then it is just 1 token per 100 versions.

Reply:
I know it has to be done by a user but that is not excuse to allow the player to get rate limited many mods rely on the GitHub API including SkyHanni so if you get them rate limited needlessly then for an hour they cant use the GitHub API which will cause many problems for several mods. 

People already get rate limited already due to how many mods use the GitHub API literally while trying to test this I got rate limited on the first request.